internal/catalogmetadata/cache/cache.go

```diff
@@ -0,0 +1,143 @@
+
+	"github.com/operator-framework/operator-controller/internal/catalogmetadata/client"
+)
+
+var _ client.Fetcher = &filesystemCache{}
+
+// NewFilesystemCache returns a client.Fetcher implementation that uses a
+// local filesystem to cache Catalog contents. When fetching the Catalog contents
+// it will:
+// - Check if the Catalog is cached
+//   - IF !cached it will fetch from the catalogd HTTP server and cache the response
+//   - IF cached it will verify the cache is up to date. If it is up to date it will return
+//     the cached contents, if not it will fetch the new contents from the catalogd HTTP
+//     server and update the cached contents.
+func NewFilesystemCache(cachePath string, tripper http.RoundTripper) client.Fetcher {
+	return &filesystemCache{
+		cachePath:              cachePath,
+		mutex:                  sync.RWMutex{},
+		tripper:                tripper,
+		cacheDataByCatalogName: map[string]cacheData{},
+	}
+}
+
+// cacheData holds information about a catalog
+// other than it's contents that is used for
+// making decisions on when to attempt to refresh
+// the cache.
+type cacheData struct {
+	ResolvedRef string
+}
+
+// FilesystemCache is a cache that
+// uses the local filesystem for caching
+// catalog contents. It will fetch catalog
+// contents if the catalog does not already
+// exist in the cache.
+type filesystemCache struct {
+	mutex                  sync.RWMutex
+	cachePath              string
+	tripper                http.RoundTripper
+	cacheDataByCatalogName map[string]cacheData
+}
+
+// FetchCatalogContents implements the client.Fetcher interface and
+// will fetch the contents for the provided Catalog from the filesystem.
+// If the provided Catalog has not yet been cached, it will make a GET
+// request to the Catalogd HTTP server to get the Catalog contents and cache
+// them. The cache will be updated automatically if a Catalog is noticed to
+// have a different resolved image reference.
+// The Catalog provided to this function is expected to:
+// - Be non-nil
+// - Have a non-nil Catalog.Status.ResolvedSource.Image
+// This ensures that we are only attempting to fetch catalog contents for Catalog
+// resources that have been successfully reconciled, unpacked, and are being served.
+// These requirements help ensure that we can rely on status conditions to determine
+// when to issue a request to update the cached Catalog contents.
+func (fsc *filesystemCache) FetchCatalogContents(ctx context.Context, catalog *catalogd.Catalog) (io.ReadCloser, error) {
+	if catalog == nil {
+		return nil, fmt.Errorf("error: provided catalog must be non-nil")
+	}
+
+	if catalog.Status.ResolvedSource == nil {
+		return nil, fmt.Errorf("error: catalog %q has a nil status.resolvedSource value", catalog.Name)
+	}
+
+	if catalog.Status.ResolvedSource.Image == nil {
+		return nil, fmt.Errorf("error: catalog %q has a nil status.resolvedSource.image value", catalog.Name)
+	}
```

Comment:
I know this may not actually be possible, but in theory shouldn't we be able to fetch catalog contents without a resolved source if we have the content URL?

It seems like these fields not being present means that we could fall back to a `If-Modified-Since` request.

Reply:
In what real-world scenario would that manifest?

Reply:
In theory yes, practically though I'm not sure that scenario would take place since the content URL is a side effect of successfully resolving and unpacking a source. 

I could be wrong, but I don't see a real world case where this happens.

Reply:
>> I know this may not actually be possible

> I could be wrong, but I don't see a real world case where this happens.

Very well!